Add support for integrated sensor pixel field visualization

Alexboiboi opened the pull request:
## Notes

This PR aims to add a new field visualization method that easily integrates with Magpylib. The `Sensor` object with its `pixel` is the perfect object to use for this purpose!
With this new capability one can display the vector field for `"B", "H", "M", or "J"` and overlay a magnitude (e.g. "Bxy", "Hz", etc.. )

Below some example on how it may be used:

### Animated B-field

![Animation](https://github.com/magpylib/magpylib/assets/29252289/96dbd80f-12b4-401f-a742-eb7bdc0cc382)
<details>
<summary>Code to reproduce</summary>

```python
import magpylib as magpy
import numpy as np

pix_per_dim = 9

c1 = magpy.magnet.Cuboid(polarization=(0, 0, 1), dimension=(1, 1, 1), style_opacity=0.2)
c1.rotate_from_angax(np.linspace(0, 180, path_len), "z", start=0)
c1.rotate_from_angax(np.linspace(0, 180, path_len), "x", start=0)

ls = np.linspace(-1, 1, pix_per_dim)
s1 = magpy.Sensor(pixel=[[x, y, 0] for x in ls for y in ls], position=(0, 0, 0))

magpy.show(
    c1,
    s1,
    animation=True,
    style_path_frames=[-1],
    style_path_show=False,
    style_pixel_field_symbol="arrow3d",
    style_pixel_field_vectorsource="B",
    style_pixel_field_colorsource="Bz",
)
```
</details>

### Other fields
![image](https://github.com/magpylib/magpylib/assets/29252289/4ffca734-49a8-4159-b4ce-06d31730e854)

<details>
<summary>Code to reproduce</summary>

```python
import magpylib as magpy
import numpy as np

path_len = 51
pix_per_dim = 9

c1 = magpy.magnet.Cuboid(polarization=(0, 0, 1), dimension=(1, 1, 1), style_opacity=0.2)
ls = np.linspace(-1, 1, pix_per_dim)
s0 = magpy.Sensor(pixel=[[x, y, 0] for x in ls for y in ls], position=(0, 0, 0))
objects = []
for i, field in enumerate("BHJ"):
    s = s0.copy(
        style_label=f"Sensor (pixel field {field!r})",
        style_pixel_field_vectorsource=field,
        style_pixel_field_colorsource=field,
    )
    objects.append({"objects": (c1, s), "col": i + 1})

magpy.show(
    *objects,
    style_path_frames=[-1],
    style_path_show=False,
    style_pixel_field_symbol="arrow3d",
)
```

</details>

Alexboiboi commented:
@OrtnerMichael, @fslanovc @pribossek would be nice to get some feedback on this one 😉 🙏

OrtnerMichael commented:
@pribossek have a look at this!

Alex found a way to incorporate a feature that you asked for several years back ...

OrtnerMichael commented:
@Alexboiboi 

A legend for the magnetic field amplitude would be really useful.

OrtnerMichael commented:
@Alexboiboi 

found a bug when pixel are unordered - I guess it has to do with the order they are displayed in.

when its ordered (line 11 deactivated) I get this

![image](https://github.com/magpylib/magpylib/assets/47384012/5f9e6d2c-9d4e-41d7-9acf-2ee2d35326ea)

when the pixle array is shuffled I get this

![image](https://github.com/magpylib/magpylib/assets/47384012/d7f4223c-3dcd-435f-94c7-3b9c4a50a597)

<details>
this is the code
```python
import magpylib as magpy
import numpy as np

cube = magpy.magnet.Cuboid(
    polarization=(0, 0, 1),
    dimension=(1,1,1),
    style_opacity=0.2,
)

pixx = np.linspace((-2,0,1), (2,0,1), 20)
np.random.shuffle(pixx)
field = magpy.Sensor(pixel=pixx)

magpy.show(
    field,
    cube,
    style_pixel_field_symbol="arrow3d",
    style_pixel_field_vectorsource="B",
)
```
<\details>

OrtnerMichael commented:
**Feature Discussion**

Instead of (ab)using the `Sensor` class for the purpose of displaying the field - why not create a new class, e.g. `FieldViewer`, - that works quite similarly but is dedicated to this display task ?

E.g. one could associate it with sources. That would make it e.g. possible to display sources that do not contribute to the fieldViewer field. Or one could use different methods of display.

OrtnerMichael commented:
Here is another **maybe this is a bug?**

This code
```python
import magpylib as magpy

coll = magpy.Collection()
for i in range(3):
    cube = magpy.magnet.Cuboid(
        position=(i,0,0),
        dimension=(1,1,1),
        polarization=(0,0,(-1)**i),
    )
    coll.add(cube)

for i,m in enumerate(coll):
    if i%2==0:
        print("tick")
        m.style.magnetization.color.north = "black"

magpy.show(coll, backend="plotly")
```
produces the following image using the **main banch**

![image](https://github.com/magpylib/magpylib/assets/47384012/1b7db0f7-f4a3-47d8-9049-b6e09d13fb66)

With the **show_sensor_pixel_field** branch the output becomes

![image](https://github.com/magpylib/magpylib/assets/47384012/00169324-b5f3-4ec6-82fd-47f501ffeb2a)

Alexboiboi commented:
@OrtnerMichael 

Nice bug catches

> found a bug when pixel are unordered - I guess it has to do with the order they are displayed in.

fixed in [2fecc96](https://github.com/magpylib/magpylib/pull/793/commits/2fecc963c0500e8204b57f481fcc349b0ae64ccc)


[565266d](https://github.com/magpylib/magpylib/pull/793/commits/565266dcf586ecf636030bcf4344d69448743f6d)

> Trying around a bit it looks like your new code sets properties somehow delayed and somehow the last style command is applied to all `cube` objects.

fixed in [565266d](https://github.com/magpylib/magpylib/pull/793/commits/565266dcf586ecf636030bcf4344d69448743f6d)

fslanovc commented:
@Alexboiboi This feature looks amazing - great to visualize the exact field profile at the sensor! 👍

Alexboiboi commented:
@OrtnerMichael  now ready to be merge

See the docs example here:
https://magpylib--793.org.readthedocs.build/en/793/_pages/user_guide/examples/examples_vis_vectorfield.html#examples-vis-vectorfield

OrtnerMichael commented:
I'll get to it asap ... big promise :)

OrtnerMichael commented:
- [ ] shownull --> showzero or showinvalid

Can we rename the kwarg `shownull` as `showzero` ?
null vector is an uncommon term, and can be confused with invalid vectors - zero vector is more common to denote (0,0,0).

OrtnerMichael commented:
Finally working on it now. I don't know why I was opposed to this - it looks really great.

- [ ] I suggest to give the default value `colorsource = "vectorsource"` instead of `colorsource=None`. This seems much more intuitive to me. `None` just confuses me - I know its usually our default.

OrtnerMichael commented:
one more problem with the outlined proposal: if you want to display a pixel field with colored cubes, those cubes will be oriented automatically along the field, which you probably do not want.

So for some 3D primitives you want to see color but not necessarily orientation.

messy messy

OrtnerMichael commented:
**After a long discussion with CoPilot**

Its mostly telling me what it thinks I want to hear, and not offering so much objective perspective - but it helped me go through the use cases. I suggest the following interface:

**`style.sensor.pixel.field=None`**
 - allow `None`, `"B"`, `"Hxy"`, `"Hx"`, etc..
 - If `None` all pixels are displayed the same (**uniform mode**), else we are in the **variable mode** where all pixel are displayed trying to visualize the given field

**`style.sensor.pixel.symbol="auto"`** - sets the pixel shape
- allow 2D marker like `"o"`, `"*"`, ... and 3D primitives like "`box`", "`cone`", ...
- in uniform mode:
    - Default (`"auto"`) is `"box"`
- in variable mode:
    - Default (`"auto"`) is `"arrow3d"`
    - Note: some symbols can display orientation, some cannot

**`style.sensor.pixel.size=1.0`**
- allow all numbers >0.
- Multiplicative factor on default pixel size which is not always optimal

**`style.sensor.pixel.sizemode="uniform"`**
 - options: `"uniform"`, `"linear"`, `"log"`, `"loglog"`
 - as already implemented

**`style.sensor.pixel.color='k'`**
 - allow any valid color `"k"`, `(.1,.2,.3)`, ...
 - sets the uniform mode pixel color

**`style.sensor.pixel.cmap='Viridis'`**
 - allow standard color maps `"Jet"`, `"RgYnBl"`, ... 
 - sets the variable mode pixel color

**`style.sensor.pixel.colormode="linear"`**
 - options: `"uniform"`, `"linear"`, `"log"`, `"loglog"`
 - when `"uniform"` display pixel in `style.sensor.pixel.color`
 
**`style.sensor.pixel.orientation="auto"`**
- options: `"field"`, `"None"`
- in uniform mode:
    - Default (`"auto"`) is `None`
    - the orientation of the primitive is not modified
- in variable mode
    - Default (`"auto"`) is `"field"`
    - the orientation of the primitive follows the orientation of the field
